storage/innobase/handler/ha_innodb.cc

```diff
@@ -16078,6 +16147,23 @@ ha_innobase::external_lock(
 	trx_t* trx = m_prebuilt->trx;
 	ut_ad(m_prebuilt->table);
 
+	bool not_started = false;
+	switch (trx->state) {
+	default:
+	case TRX_STATE_PREPARED:
+		ut_ad("invalid state" == 0);
+		DBUG_RETURN(HA_ERR_WRONG_COMMAND);
+	case TRX_STATE_ABORTED:
+		/* Reset the state if the transaction had been aborted. */
+		trx->state = TRX_STATE_NOT_STARTED;
```

Comment:
This logic was in the original patch by @montywi. If I understand correctly, this is intended only for `autocommit=1` transactions. In multi-statement transactions, we’d probably better return an error instead.

Reply:
This was addressed already in 41839d2d3d4af71a12336b108fda2f54098828c5.